### PR TITLE
Typefy Editor Datetimes Components closes #2236

### LINKED ISSUE #2236: Convert Event Editor Datetimes Components to TypeScript
- [ ] domain/eventEditor/datetimes/DatetimeId.js
- [ ] domain/eventEditor/datetimes/dateCard/*
- [ ] domain/eventEditor/datetimes/dateForm/*
- [ ] domain/eventEditor/datetimes/dateList/AddNewDateModal.js
- [ ] domain/eventEditor/datetimes/dateList/filterBar/*

`*` === "everything in folder"

### PR DESCRIPTION
**Pull request by manzoorwanijk**
This PR:
- Converts `/domain/eventEditor/datetimes` components to TS.
- Converts the dependencies to TS.
- Closes #2236

### CODE REVIEW

File: assets/prototype/application/services/apollo/mutations/index.ts [hunk 1]
```diff
@@ -1,3 +1,3 @@
 export { default as useEntityMutator } from './useEntityMutator';
 
-export { EntityType } from './types';
+export * from './types';
```
Comment (nicolad) [inline assets/prototype/application/services/apollo/mutations/index.ts]: :+1:

File: assets/prototype/application/services/apollo/mutations/useEntityMutator.ts [hunk 2]
```diff
@@ -12,7 +12,7 @@ import {
  * @param {string} type Entity type name
  * @param {string} id   Entity id
  */
-const useEntityMutator = (type: EntityType, id?: string): EntityMutator => {
+const useEntityMutator = (type: EntityType, id = ''): EntityMutator => {
```
Comment (nicolad) [inline assets/prototype/application/services/apollo/mutations/useEntityMutator.ts]: It's better empty string as default than optional :+1:

File: assets/prototype/domain/eventEditor/context/DateTimeProvider/index.tsx [hunk 3]
```diff
@@ -9,15 +9,16 @@ import {
...
 } from '../../../../application/ui/components/layout/editor-modal/useEditorModalState/types';
 
-export interface DateTimeContextProps extends ContextProps {}
-export const DateTimeContext = createContext({} as DateTimeContextProps);
+export type DateTimeContextProps = ContextProps;
+export const DateTimeContext = createContext<DateTimeContextProps>(null);
 
 interface DatetimeProviderProps {
-	children: React.ReactChildren;
+	children?: React.ReactNode;
```
Comment (nicolad) [inline assets/prototype/domain/eventEditor/context/DateTimeProvider/index.tsx]: Why this needs to be optional?

File: assets/prototype/domain/eventEditor/datetimes/DatesList/filterBar/controls/DatesSortedByControl.tsx [hunk 4]
```diff
@@ -1,18 +1,26 @@
+import React, { useMemo } from 'react';
```
Comment (nicolad) [inline assets/prototype/domain/eventEditor/datetimes/DatesList/filterBar/controls/DatesSortedByControl.tsx]: This is not an external import?

Comment (nicolad) [reply assets/prototype/domain/eventEditor/datetimes/DatesList/filterBar/controls/DatesSortedByControl.tsx]: It should be below, under:
```
/**
 * External imports
 */
```

File: assets/prototype/domain/eventEditor/datetimes/DatesList/filterBar/useDatesListFilterState/index.ts [hunk 5]
```diff
@@ -38,21 +41,21 @@ const useDatesListFilterState = (dates: Datetime[]): DatesListFilterState => {
...
-	const setDisplayDates = (displayDates: DisplayDates) => {
+	const setDisplayDates = (displayDates: DisplayDates): void => {
 		dispatch({
 			type: ActionType.SET_DISPLAY_DATES,
 			displayDates,
 		});
 	};
 
-	const setShowDates = (showDates: ShowDates) => {
+	const setShowDates = (showDates: ShowDates): void => {
```
Comment (nicolad) [inline assets/prototype/domain/eventEditor/datetimes/DatesList/filterBar/useDatesListFilterState/index.ts]: Nice :+1:

File: assets/prototype/domain/eventEditor/datetimes/DatesList/AddNewDateModal.tsx [hunk 6]
```diff
@@ -1,12 +1,17 @@
+import React from 'react';
 import DateForm from '../dateForm/DateForm';
 import FormModal from '../../../../application/ui/components/forms/FormModal';
 import { useEntityMutator, EntityType } from '../../../../application/services/apollo/mutations';
 import useTickets from '../../data/queries/tickets/useTickets';
+import { AddItemModalProps } from '../../types';
+import { DateItemFormProps } from '../types';
 
-const AddNewDateModal = ({ onClose, isOpen }) => {
+const AddNewDateModal: React.FC<AddItemModalProps> = ({ onClose, isOpen }): JSX.Element => {
```
Comment (tn3rb) [inline assets/prototype/domain/eventEditor/datetimes/DatesList/AddNewDateModal.tsx]: DUDE!!! thought you said in chat you preferred `FunctionComponent ` !!! 😆 

I'm ok with using `FC` as long as we do it consistently

File: assets/prototype/domain/eventEditor/datetimes/dateCard/DateCard.tsx [hunk 7]
```diff
@@ -51,17 +34,17 @@ const DateCard = ({ id }) => {
...
 	});
 
-	const startDate = moment(date.startDate).toDate() || PLUS_ONE_MONTH;
-	const endDate = moment(date.endDate).toDate() || PLUS_TWO_MONTHS;
-	const defaultRangeValues = [startDate, endDate];
-	const [range, setRange] = useState(defaultRangeValues);
+	const startDate: Date = moment(date.startDate).toDate() || PLUS_ONE_MONTH;
+	const endDate: Date = moment(date.endDate as moment.MomentInput).toDate() || PLUS_TWO_MONTHS;
+	const defaultRangeValues: [Date, Date] = [startDate, endDate];
+	const [range, setRange] = useState<[Date, Date]>(defaultRangeValues);
```
Comment (nicolad) [inline assets/prototype/domain/eventEditor/datetimes/dateCard/DateCard.tsx]: Great job :taco:

File: assets/prototype/domain/eventEditor/context/DateTimeProvider/index.tsx (hunk 3 above)
Comment (manzoorwanijk) [reply assets/prototype/domain/eventEditor/context/DateTimeProvider/index.tsx]: Otherwise you will have to pass `children` as prop 😄

File: assets/prototype/domain/eventEditor/datetimes/DatesList/filterBar/controls/DatesSortedByControl.tsx (hunk 4 above)
Comment (manzoorwanijk) [reply assets/prototype/domain/eventEditor/datetimes/DatesList/filterBar/controls/DatesSortedByControl.tsx]: Yes, it is

Comment (tn3rb) [reply assets/prototype/domain/eventEditor/datetimes/DatesList/filterBar/controls/DatesSortedByControl.tsx]: AFAIC we don't need to add those comment blocks in to identify External vs Internal imports.

I **DO** think we should separate them though with a single line space and put them in alphabetical order based on the name of the lib.

Those comment blocks like Vadim posted above were something Darren wanted because that's what WP core did in the GB code.

File: assets/prototype/domain/shared/dateRangeInput/formatSelect.tsx [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+const momentFormatter = (format: string): MomentFormatter => ({
+	formatDate: (date: Date): string => moment(date).format(format),
+	parseDate: (str: string): Date => moment(str, format).toDate(),
+	placeholder: `${format} (moment)`,
+});
+
+export const FORMATS = {
+	MYSQL: momentFormatter('YYYY-MM-DD HH:mm:ss'),
+};
```
Comment (nicolad) [inline assets/prototype/domain/shared/dateRangeInput/formatSelect.tsx]: The original code was written before we decided to go with `date-fns`, not sure if we want or not to keep moment along.
This is outside of the scope of the current PR.

Comment (manzoorwanijk) [reply assets/prototype/domain/shared/dateRangeInput/formatSelect.tsx]: Yes, it caused errors at in its consumer components.

File: assets/prototype/domain/eventEditor/datetimes/DatesList/filterBar/useDatesListFilterState/index.ts [hunk 9]
```diff
@@ -22,14 +22,17 @@ interface DatesListFilterState {
...
 	processedDates: Datetime[];
+	showDates: ShowDates;
+}
+
+interface DatesListFilterStateManager extends DatesListFilterState {
 	setDatesSortedBy: (datesSortedBy: DatesSortedBy) => void;
 	setDisplayDates: (displayDates: DisplayDates) => void;
 	setShowDates: (showDates: ShowDates) => void;
-	showDates: ShowDates;
 }
```
Comment (tn3rb) [inline assets/prototype/domain/eventEditor/datetimes/DatesList/filterBar/useDatesListFilterState/index.ts]: 👍

File: assets/prototype/domain/eventEditor/datetimes/dateCard/DateCard.tsx (hunk 7 above)
Comment (tn3rb) [reply assets/prototype/domain/eventEditor/datetimes/dateCard/DateCard.tsx]: agreed...
 hey  uh...
  where'd you get that taco from?

File: assets/prototype/domain/eventEditor/datetimes/dateCard/EditDateModal.tsx [hunk 10]
```diff
@@ -1,20 +1,21 @@
...
-	const onClose = () => editorState.closeEditor(editors.editForm);
+	const onClose = (): void => editorState.closeEditor(editors.editForm);
 
 	const { updateEntity } = useEntityMutator(EntityType.Datetime, id);
-	const onSubmit = (fields) => updateEntity(fields);
+	const onSubmit = (fields: any): MutationResult => updateEntity(fields);
 	const tickets = useTickets();
 
-	const formComponent = (props) => (
+	const formComponent: React.FC<any> = (props): JSX.Element => (
```
Comment (tn3rb) [inline assets/prototype/domain/eventEditor/datetimes/dateCard/EditDateModal.tsx]: is there a built-in interface for props in React?

if not maybe we should create on like:

```ts
interface componentProps {
	[ key: string ]: any
}
```

just to denote that the props at least have to be an object.

thoughts?

File: assets/prototype/domain/eventEditor/types.ts [hunk 11]
```diff
@@ -29,3 +29,30 @@ declare global {
...
+
+export interface AddItemModalProps {
+	onClose: () => void;
+	isOpen: boolean;
+}
+
+export interface EditItemFormProps {
+	title?: string;
+	formReset?: boolean;
+}
```
Comment (tn3rb) [inline assets/prototype/domain/eventEditor/types.ts]: very nice! love it 💯

File: assets/prototype/domain/shared/dateRangeInput/formatSelect.tsx (hunk 8 above)
Comment (tn3rb) [reply assets/prototype/domain/shared/dateRangeInput/formatSelect.tsx]: created #2255 to deal with that

File: assets/prototype/domain/eventEditor/datetimes/DatesList/AddNewDateModal.tsx (hunk 6 above)
Comment (manzoorwanijk) [reply assets/prototype/domain/eventEditor/datetimes/DatesList/AddNewDateModal.tsx]: VS Code was lazy to suggest and I was lazier to type it 😄

File: assets/prototype/domain/eventEditor/datetimes/DatesList/filterBar/controls/DatesSortedByControl.tsx (hunk 4 above)
Comment (nicolad) [reply assets/prototype/domain/eventEditor/datetimes/DatesList/filterBar/controls/DatesSortedByControl.tsx]: I am ok without those comments, I thought it is some kind of unwritten convention on the project.

File: assets/prototype/domain/eventEditor/datetimes/dateCard/DateCard.tsx (hunk 7 above)
Comment (nicolad) [reply assets/prototype/domain/eventEditor/datetimes/dateCard/DateCard.tsx]: we had used tacos to praise peeps when they are doing great things on the previous project: https://www.heytaco.chat/
